### PR TITLE
[SDK-3331] Spring boot Webapp QS migration

### PR DESCRIPTION
**Pull request by poovamraj**
We have migrated the Spring webapp Quickstart to the interactive version. Added screenshots of the content and code

image

image

### CODE REVIEW

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+::: panel Using Spring WebFlux?
+This tutorial uses [Spring MVC](https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html). If you are using [Spring WebFlux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#spring-web-reactive), the steps to add authentication are similar, but some of the implementation details are different. Refer to the [Spring Boot WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-boot-login-samples/tree/master/webflux-login) to see how to integrate Auth0 with your Spring Boot WebFlux application.
+:::
+
+<%= include('../../_includes/_configure_auth0_interactive', { 
+  callback: 'http://localhost:3000/login/oauth2/code/auth0',
+  returnTo: 'http://localhost:3000'
+}) %>
+
+## Configure Spring Boot Application
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Configure Spring Boot application
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+Spring Security makes it easy to configure your application for authentication with OIDC providers such as Auth0. In your application's configuration, configure the OAuth2 client and provider. The sample to the right shows an `application.yml` file, though you can also use properties files or any of the other [supported externalization mechanisms](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config).
+
+:::note
+Spring Security will use the `issuer-uri` property value to retrieve all the information necessary to enable login and validation of the ID token at runtime.
+
+[Additional property mappings](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2login-boot-property-mappings) are available for further customization, if required.
+:::
+
+## Add Login to Your Application {{{ data-action=code data-code="SecurityConfig.java" }}}
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Add login to your application {{{ data-action=code data-code="SecurityConfig.java" }}}
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+```java
+http.authorizeRequests()
+      .mvcMatchers("/").permitAll()
+      .anyRequest().authenticated()
+      .and().oauth2Login();
+```
+:::
+
+## Add Front Page {{{ data-action=code data-code="index.html" }}}
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Add front page {{{ data-action=code data-code="index.html" }}}
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+Spring Security will use the client configuration you defined earlier to handle login when a user visits the `/oauth2/authorization/auth0` path of your application. You can use this to create a login link in your application.
+
+We show the user attributes in this page when the user is authenticated. The `/logout` link in the template will be later used while implementing the Logout feature. 
+
+## Add Controller {{{ data-action=code data-code="HomeController.java" }}}
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Add controller {{{ data-action=code data-code="HomeController.java" }}}
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+::::
+
+![Auth0 Universal Login](/media/quickstarts/universal-login.png)
+
+:::note
+Auth0 enables the Google social provider by default on new tenants and offers you developer keys to test logging in with [social identity providers](https://auth0.com/docs/connections/identity-providers-social). However, these developer keys have some limitations that may cause your application to behave differently. For more details on what this behavior may look like and how to fix it, consult the [Test Social Connections with Auth0 Developer Keys](https://auth0.com/docs/connections/social/devkeys#limitations-of-developer-keys) document.
+:::
+
+## Add Logout to Your Application {{{ data-action=code data-code="LogoutHandler.java" }}}
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Add logout to your application {{{ data-action=code data-code="LogoutHandler.java" }}}
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+Now that users can log into your application, they need [a way to log out](https://auth0.com/docs/logout/guides/logout-auth0). By default, when logout is enabled, Spring Security will log the user out of your application and clear the session. To enable successful logout of Auth0, you can extend the `SecurityContextLogoutHandler` class to redirect users to your [Auth0 logout endpoint](https://auth0.com/docs/api/authentication?javascript#logout) (`https://${account.namespace}/v2/logout`) and then immediately redirect them to your application.
+
+## Update your Security Configuration {{{ data-action=code data-code="SecurityConfigWithLogout.java" }}}
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
## Update your security configuration {{{ data-action=code data-code="SecurityConfigWithLogout.java" }}}
```

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  - files/logout-handler
+  - files/security-config-logout
+github:
+    path: mvc-login
+---
+
+# Add Login to your Spring Webapp
+
+::: panel Using Spring WebFlux?
+This tutorial uses [Spring MVC](https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html). If you are using [Spring WebFlux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#spring-web-reactive), the steps to add authentication are similar, but some of the implementation details are different. Refer to the [Spring Boot WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-boot-login-samples/tree/master/webflux-login) to see how to integrate Auth0 with your Spring Boot WebFlux application.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
This tutorial uses [Spring MVC](https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html). If you are using [Spring WebFlux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#spring-web-reactive), the steps to add authentication are similar, but some implementation details are different. Refer to the [Spring Boot WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-boot-login-samples/tree/master/webflux-login) to see how to integrate Auth0 with your Spring Boot WebFlux application.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    </dependency>
+    <dependency>
+        <groupId>org.thymeleaf.extras</groupId>
+        <artifactId>thymeleaf-extras-springsecurity5</artifactId>
+    </dependency>
+</dependencies>
+```
+
+:::note
+The Spring Security 5.4.0 release will include [a fix](https://github.com/spring-projects/spring-security/pull/8357) to validate the issuer claim of the ID token. You should update to this release when available.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
The Spring Security 5.4.0 release includes [a fix](https://github.com/spring-projects/spring-security/pull/8357) to validate the ID token issuer claim. Upgrade to 5.4.0 or higher when possible.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+:::
+
+## Configure Spring Security {{{ data-action=code data-code="application.yml#1:16" }}}
+
+Spring Security makes it easy to configure your application for authentication with OIDC providers such as Auth0. In your application's configuration, configure the OAuth2 client and provider. The sample to the right shows an `application.yml` file, though you can also use properties files or any of the other [supported externalization mechanisms](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config).
+
+:::note
+Spring Security will use the `issuer-uri` property value to retrieve all the information necessary to enable login and validation of the ID token at runtime.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Spring Security uses the `issuer-uri` property value to retrieve the information necessary to enable login and ID token validation at runtime.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+[Additional property mappings](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2login-boot-property-mappings) are available for further customization, if required.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
If you need more property mappings, [review the Spring documentation](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2login-boot-property-mappings) for further customization.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+:::
+
+## Add login to your application {{{ data-action=code data-code="SecurityConfig.java" }}}
+
+To enable users to login with Auth0, you will need to extend the [WebSecurityConfigurerAdapter](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/configuration/WebSecurityConfigurerAdapter.html) class and override the `configure(HttpSecurity http)` method.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
To enable user login with Auth0, extend the [WebSecurityConfigurerAdapter](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/configuration/WebSecurityConfigurerAdapter.html) class and override the `configure(HttpSecurity http)` method.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+Later in this quickstart we will overwrite this file with `SecurityConfigWithLogout.java` to provide extra configurations to support the Logout feature.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Later in this quickstart, you will overwrite this file with `SecurityConfigWithLogout.java` to provide extra configurations to support the logout feature.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+:::note
+You can further configure the [HttpSecurity](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html) instance to require authentication on all or certain paths. For example, to require authentication on all paths except the home page:
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
You can configure the [HttpSecurity](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html) instance to require authentication on all or certain paths. For example, to require authentication on all paths except the home page:
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+http.authorizeRequests()
+      .mvcMatchers("/").permitAll()
+      .anyRequest().authenticated()
+      .and().oauth2Login();
+```
+:::
+
+## Add front page {{{ data-action=code data-code="index.html" }}}
+
+Spring Security will use the client configuration you defined earlier to handle login when a user visits the `/oauth2/authorization/auth0` path of your application. You can use this to create a login link in your application.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Spring Security uses the client configuration you defined earlier to handle login when a user visits the `/oauth2/authorization/auth0` path of your application. You can use this to create a login link in your application.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+We show the user attributes in this page when the user is authenticated. The `/logout` link in the template will be later used while implementing the Logout feature.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
This page returns the user attributes when the user authentications. You will use the `/logout` link in the template to implement the logout feature.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+
+## Add front page {{{ data-action=code data-code="index.html" }}}
+
+Spring Security will use the client configuration you defined earlier to handle login when a user visits the `/oauth2/authorization/auth0` path of your application. You can use this to create a login link in your application.
+
+We show the user attributes in this page when the user is authenticated. The `/logout` link in the template will be later used while implementing the Logout feature. 
+
+## Add controller {{{ data-action=code data-code="HomeController.java" }}}
+
+We will create a controller to handle the incoming request. This controller will be used to render our `index.html` page. When the user is authenticated, we will setup the attributes user's profile information that will be used while rendering page.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Create a controller to handle the incoming request. This controller renders the `index.html` page. When the user authenticates, the application retrieves the users's profile information attributes to render the page.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+
+::::checkpoint
+
+:::checkpoint-default
+
+When you click the login link, verify that your application redirects you to the [Auth0 Universal Login](https://auth0.com/universal-login) page and that you can now log in or sign up using a username and password or a social provider.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
When you click the login link, verify the application redirects you to the [Auth0 Universal Login](https://auth0.com/universal-login) page and that you can now log in or sign up using a username and password or a social provider.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+:::
+
+:::checkpoint-failure
+Sorry about that. Here's a couple things to double check:
+* you configured the correct Callback URL
+* you added the login link to redirect to `/oauth2/authorization/auth0`
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
If your application did not allow login or signup:
* Verify you configured the correct Callback URL
* Verify you added the login link to redirect to `/oauth2/authorization/auth0`
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+Auth0 enables the Google social provider by default on new tenants and offers you developer keys to test logging in with [social identity providers](https://auth0.com/docs/connections/identity-providers-social). However, these developer keys have some limitations that may cause your application to behave differently. For more details on what this behavior may look like and how to fix it, consult the [Test Social Connections with Auth0 Developer Keys](https://auth0.com/docs/connections/social/devkeys#limitations-of-developer-keys) document.
+:::
+
+## Add logout to your application {{{ data-action=code data-code="LogoutHandler.java" }}}
+
+Now that users can log into your application, they need [a way to log out](https://auth0.com/docs/logout/guides/logout-auth0). By default, when logout is enabled, Spring Security will log the user out of your application and clear the session. To enable successful logout of Auth0, you can extend the `SecurityContextLogoutHandler` class to redirect users to your [Auth0 logout endpoint](https://auth0.com/docs/api/authentication?javascript#logout) (`https://${account.namespace}/v2/logout`) and then immediately redirect them to your application.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Now that users can log into your application, they need [a way to log out](https://auth0.com/docs/logout/guides/logout-auth0). By default, Spring Security logs user out of your application and clears the session when you enable logout. To enable successful Auth0 logout, extend the `SecurityContextLogoutHandler` class to redirect users to your [Auth0 logout endpoint](https://auth0.com/docs/api/authentication?javascript#logout) (`https://${account.namespace}/v2/logout`) and then immediately redirect them to your application.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+
+## Update your security configuration {{{ data-action=code data-code="SecurityConfigWithLogout.java" }}}
+
+Next, you need to update your implementation of `WebSecurityConfigurerAdapter` to register your logout handler and specify the request path that should trigger logout (`/logout` in the example below).
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
Next, update your implementation of `WebSecurityConfigurerAdapter` to register your logout handler and specify the request path that should trigger logout (`/logout` in the example below).
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+
+You can remove the `SecurityConfig.java` and replace it with `SecurityConfigWithLogout.java` or just update the contents from the one file to another.
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
You can remove the `SecurityConfig.java` and replace it with `SecurityConfigWithLogout.java` or update the contents from the one file to another.
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+::::checkpoint
+
+:::checkpoint-default
+When you click logout link, the application should redirect you to the address you specified as one of the "Allowed Logout URLs" in the "Settings" and you are no longer logged in to your application.
+:::
+
+:::checkpoint-failure
+Sorry about that. Here's a couple things to double check:
+* you configured the correct Logout URL
+* you added the Logout link to redirect to `/logout`
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
If your application did not allow logout:
* Verify you configured the correct logout URL
* Verify you added the logout link to redirect to `/logout`
```

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+  - files/security-config
+  - files/index
+  - files/home-controller
+  - files/logout-handler
+  - files/security-config-logout
+github:
+    path: mvc-login
+---
+
+# Add Login to your Spring Webapp
```
Comment (tr0njavolta) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
# Add login to your Spring Webapp
```

**Review by tr0njavolta — COMMENTED** [review]
Some style updates

File: articles/quickstart/webapp/java-spring-boot/interactive.md [hunk 24]
```diff
@@ -0,0 +1,175 @@
+---
+title: Add Login to your Spring Webapp
```
Comment (frederikprijck) [inline articles/quickstart/webapp/java-spring-boot/interactive.md]:
```suggestion
title: Add login to your Spring Webapp
```

**Review by frederikprijck — CHANGES_REQUESTED** [review]

**Review by frederikprijck — APPROVED** [review]

**Review by tr0njavolta — APPROVED** [review]

**Pull request merged**